nupic/research/frameworks/greedy_infomax/models/BilinearInfo.py

```diff
@@ -0,0 +1,198 @@
+                          For example, if k_predictions is 5 and skip_step is 1,
+                          then this module will compare z_{t} with z_{t+2}...z{t+6}.
+
+    """
+
+    def __init__(self, in_channels, out_channels, negative_samples=16, k_predictions=5):
+        super().__init__()
+        self.negative_samples = negative_samples
+        self.k_predictions = k_predictions
+
+        # 1x1 convolutions to sample from the channel dimension
+        # representations converted from in_channels to out_channels dimensions
+        self.W_k = nn.ModuleList(
+            nn.Conv2d(in_channels, out_channels, 1, bias=False)
+            for _ in range(self.k_predictions)
+        )
+
+    def forward(self, z, c, skip_step=1):
+        """
+        :param z: positive samples z_{t} (used to create z_{t+k} for each k)
+        :param c: context (reference samples). In audio experiments, this is an
+                  autoregressive representation which captures the entire temporal
+                  history up until time t. In vision experiments, this is simply
+                  z_{t}, which is the same as parameter z.
+        :param skip_step: Number of steps to skip between the reference sample and
+                          the positive samples. Defaults to 1, as in the paper.
+                          Skipping 1 step prevents overlapping samples from being
+                          compared with the default overlap. This parameter should
+                          probably be increased for larger overlap values.
+        """
+        batch_size = z.shape[0]
+
+        # For each k in k_predictions, store a set of log_fk and true_f values
+        log_f_list, true_f_list = [], []
+        for k in range(1, self.k_predictions + 1):
+            # Compute log f(c_t, x_{t+k}) = z^T_{t+k} W_k c_t
+
+            # First half of bilinear model, compute z^T_{t+k} W_k:
+            ztwk = (
+                self.W_k[k - 1]  # 1x1 convolution
+                .forward(z[:, :, (k + skip_step) :, :])  # Bx, C , H , W
+                .permute(2, 3, 0, 1)  # H, W, Bx, C
+                .contiguous()
+            )  # y, x, b, c
+
+            # Take random samples from z_{t+k} W_k as contrastive samples
+            ztwk_shuf = ztwk.view(
+                ztwk.shape[0] * ztwk.shape[1] * ztwk.shape[2], ztwk.shape[3]
+            )  # y * x * batch, c
+            rand_index = torch.randint(
+                ztwk_shuf.shape[0],  # y *  x * batch
+                (ztwk_shuf.shape[0] * self.negative_samples, 1),
+                dtype=torch.long,
+                device=ztwk_shuf.device,
+            )
+            rand_index = rand_index.repeat(1, ztwk_shuf.shape[1])
+            ztwk_shuf = torch.gather(
+                ztwk_shuf, dim=0, index=rand_index, out=None
+            )  # y * x * b * n, c
+            ztwk_shuf = ztwk_shuf.view(
+                ztwk.shape[0],
+                ztwk.shape[1],
+                ztwk.shape[2],
```

Comment:
Style comment: this can be reduced to fewer lines, instead of having each argument on a separate line.